File: peers/node/http-client.mjs
import { Client, StreamableHTTPClientTransport } from "@modelcontextprotocol/client";

function textOf(result) {
  const content = result?.content;
  if (Array.isArray(content) && content[0]?.text) return content[0].text;
  return String(result ?? "");
}

function resourceText(result) {
  const contents = result?.contents;
  if (Array.isArray(contents) && contents[0]?.text) return contents[0].text;
  return String(result ?? "");
}

function promptText(result) {
  return JSON.stringify(result?.messages ?? result ?? "");
}

const url = process.argv[2];
if (!url) {
  console.error("usage: http-client.mjs <url>");
  process.exit(2);
}

const client = new Client({ name: "mcp-parity-node", version: "0.1.0" });
client.setVersionNegotiation({ mode: "auto" });
await client.connect(new StreamableHTTPClientTransport(new URL(url)), {
  timeout: 120_000,
});

const tools = await client.listTools();
const echo = await client.callTool({ name: "echo", arguments: { msg: "pong" } });
const resource = await client.readResource({ uri: "memo://hi" });
const prompt = await client.getPrompt({ name: "greet" });

const rec = {
  era: client.getProtocolEra?.() ?? client.getNegotiatedProtocolVersion?.() ?? null,
  tools: (tools.tools ?? []).map((t) => t.name),
  echo: textOf(echo),
  resource: resourceText(resource),
  prompt: promptText(prompt),
};
process.stdout.write(`${JSON.stringify(rec)}\n`);
await client.close?.();
process.exit(0);

File: peers/python/http_client.py
#!/usr/bin/env python3
from __future__ import annotations

import asyncio
import json
import sys

from fastmcp import Client

from probe import probe_invalid_echo


def _text(result: object) -> str:
    data = getattr(result, "data", None)
    if data is not None:
        return str(data)
    content = getattr(result, "content", None)
    if content:
        first = content[0]
        return str(getattr(first, "text", first))
    return str(result)


def _resource_text(contents: object) -> str:
    if not contents:
        return ""
    first = contents[0]
    return str(getattr(first, "text", first))


def _prompt_text(prompt: object) -> str:
    messages = getattr(prompt, "messages", prompt)
    return str(messages)


async def main() -> None:
    if len(sys.argv) < 2:
        print("usage: http_client.py <url>", file=sys.stderr)
        raise SystemExit(2)
    async with Client(sys.argv[1]) as client:
        tools = await client.list_tools()
        echo = await client.call_tool("echo", {"msg": "pong"})
        resource = await client.read_resource("memo://hi")
        prompt = await client.get_prompt("greet")
        rec = {
            "era": getattr(client, "protocol_version", None) or "legacy",
            "tools": [t.name for t in tools],
            "echo": _text(echo),
            "resource": _resource_text(resource),
            "prompt": _prompt_text(prompt),
            **(await probe_invalid_echo(client)),
        }
        print(json.dumps(rec, ensure_ascii=False), flush=True)


if __name__ == "__main__":
    asyncio.run(main())
